### ISSUE TITLE
No way to force newline between block closing brace only for specific entities

### ISSUE THREAD

**Issue by ArmorDarks** [stylelint/stylelint#2759]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Feature request.

It seems that there is no way to enforce newline between block closing brace and ruleset or directive, but disallow it for regular CSS properties.

Note, that there is an issue with directives like `@include`. We can't know, does it contain whole new ruleset or something else, and depending on it author might print it differently:

```scss
.test1 {
  border: 1px;
  
  // by looking on those newlines, we can understand that this mixins contains whole new ruleset and creates nesting
  @include test1();

}

// And here it probably simply adds few properties to `.test2`
.test2 {
  border: 1px;
  @include test2();
}
```

Because of that, ability to make newline optional for some entities will be beneficial.

> Which rule, if any, is this issue related to?

Most likely related to `block-closing-brace-empty-line-before`

> What CSS is needed to reproduce this issue?

```scss
.test {
  border: 1px;

  > .child {
    margin-top: 21px;
    margin-bottom: 21px;
  }
  // <- no way to force this newline only after ruleset, but not property
}

.test2 {
  border: 1px;

  @include test2() {
    margin-top: 21px;
    margin-bottom: 21px;
  }
  // <- no way to make newline for directives optional
}

.test3 {
  border: 1px; // <- for properties should be rule, which will disallow newline before closing brace
}
```

> Which version of stylelint are you using?

`8.0.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint *.scss`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yeap, it is related to ability of SCSS to nest rulesets

> What did you expect to happen?

Ability to have option for `block-closing-brace-empty-line-before` which will reverse behavior for rulesets, or will apply different behavior for different entities.

> What actually happened (e.g. what warnings or errors you are getting)?

No way to define different behavior for newline between block closing brace and property, and block closing brace for ruleset.

**Comment by jeddy3** [comment]
@ArmorDarks Thanks again for using the issue template and for your thorough write-ups.

> Ability to have option for block-closing-brace-empty-line-before which will reverse behavior for rulesets

I believe this is a duplicate of https://github.com/stylelint/stylelint/issues/2090. That issue has been around for a while and hasn't been picked up by anyone yet. So, feel free to [contribute the option](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#adding-an-option-to-an-existing-rule).

> Note, that there is an issue with directives like @include. We can't know, does it contain whole new ruleset or something else, and depending on it author might print it differently:

Non-standard features are handled within 3rd party plugins. For this specific use case, it might be worth [writing your own plugin](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md) that replaces the built-in `block-closing-brace-empty-line-before` rule with something to, perhaps, uses heuristics to determine whether an empty line is required or not.

**Comment by jeddy3** [comment]
Labelling as `plugin` for the second request. Feel free to continue the discussion if my triaging doesn't resolve the issue.

**Comment by ArmorDarks** [comment]
Ah, indeed it seems to be a duplicate of #2090, thanks for pointing it out. I didn't notice it.
